Proposal: improved completion results

Sarcasm opened the pull request:
First,
Thanks for cquery, this is a very promising project!

I'd like to propose some changes to the completion results.
You can find attached to this PR a draft implementation,
if you want to try it out.
I will be willing to clean it if there is interest in the changes I propose.

First, a list of the few things that I think could be improved:
- The detail, which is the full signature, repeats the name of the candidate.
  This is mostly annoying an issue when the detail is "inline" (<kbd>Ctrl + Space</kbd> in VS Code).
- In C++, it's possible to have multiple function with the same name,
  but different arguments,
  it is interesting to be able to pick up the right function at a glimpse,
  with the snippet completing the arguments.

My solutions:
- Show the contextual parent of the candidate in the details,
  this is a new information we didn't have before.
- Show the function/variable signature instead simple identifier for the label.
  This allow one to pick up the right signature directly.
- Expand default arguments into multiple items,
  so that you can pick up the right signature directly,
  the same way you do with overloaded functions.

Before/After on custom/simplified struct:

![image](https://user-images.githubusercontent.com/424046/35697445-8bb65d3a-078a-11e8-832a-ddd7474c7849.png)
![image](https://user-images.githubusercontent.com/424046/35698290-e19d2650-078c-11e8-8346-107cb07c6444.png)

Before/After with STL, heavy on the signature:

![image](https://user-images.githubusercontent.com/424046/35697719-4e5913aa-078b-11e8-9f85-860c1b283051.png)
![image](https://user-images.githubusercontent.com/424046/35698240-bfb6ecf6-078c-11e8-8852-1e4c83f7e981.png)



For reference, the boilerplate code I used for quick testing:

```cpp
#include <string>

struct Abc {
  /// A function with 2 defaults arguments.
  void defarg(int a = 0, int b = 0);

  /// This overload takes an integer.
  void overloaded(int x);

  /// This overload takes a double.
  void overloaded(double x);

  int foobiz = 42;
};

void someFn(std::string &s) {
   
}
```

What do you think?
If you are positive, I will be glad to cleanup the PR for proper submission.

MaskRay commented on third_party/string_view.h:
```diff
@@ -1,6 +1,2 @@
-#if __cplusplus >= 201703L || defined(__has_include) && __has_include(<string_view>)
```
https://github.com/cquery-project/cquery/issues/390

On Mac OS X, Xcode somehow includes `<string_view>` so these lines cannot be removed

Sarcasm replied:
I wanted to sync my repository before submitting and this issue popped up, so this was just a quickfix.
This change should be a separate issue, which I assume others will encounter soon.

jacobdufault commented:
These changes sound great. We should keep the current name/detail split as an option though. Thanks for working on this!

MaskRay commented on src/clang_complete.cc:
```diff
@@ -149,147 +149,122 @@ lsCompletionItemKind GetCompletionKind(CXCursorKind cursor_kind) {
+    }
+  }
 
-      case CXCompletionChunk_LeftParen:
```
I can clean this a bit to make you easier..

MaskRay commented:
I agree that displaying the comments of the contect (e.g. `abc.` in your example) will be useful.

![](https://ptpb.pw/GIRl.jpg)

Does it look as nice in Emacs with your PR?

Sarcasm commented:
Yeah it looks reasonable in Emacs, but I'd like to make some fixes to company-lsp once I get the feature into cquery.

It's a bit difficult because VS Code seems to be tailored for LSP, but company mode has to adapt to look as cool, but I'm sure we can get something good too!

![image](https://user-images.githubusercontent.com/424046/35703252-bd036084-079b-11e8-9215-e1de81bce7b8.png)

![image](https://user-images.githubusercontent.com/424046/35703206-a12f7852-079b-11e8-92f8-b832d7dfe026.png)


> We should keep the current name/detail split as an option though.

Ok, I will look at making this customizable then.

MaskRay commented:
`ClangCompleteManager` has a member `Config* config_` which is the initialization options. You may add a `struct Completion` option in https://github.com/cquery-project/cquery/blob/master/src/config.h#L131

Sarcasm commented:
Thanks @MaskRay for the pointer to the config, that was very helpful.
I cleaned the patch and fixed a bunch of stuff.
Now the sorting/filterting of completions is working, it wasn't right in my previous attempt.

Here are some screenshots:

**VS Code**

![image](https://user-images.githubusercontent.com/424046/35773467-e7401630-0953-11e8-9e4c-9594c2f8d689.png)
![image](https://user-images.githubusercontent.com/424046/35773479-3a64e840-0954-11e8-87a5-e04c1ea8c768.png)

**Emacs + company-lsp**

![image](https://user-images.githubusercontent.com/424046/35773454-81a0c248-0953-11e8-92eb-cf34b5cf5a40.png)
![image](https://user-images.githubusercontent.com/424046/35773461-b5ac02be-0953-11e8-90fa-b9188e8479bf.png)

MaskRay commented on src/messages/text_document_completion.cc:
```diff
@@ -68,17 +68,16 @@ struct Out_TextDocumentComplete
+  const auto lhsLabelLength = lhs.label.length();
+  const auto rhsFilterTextLength = rhs.filterText.length();
+  const auto rhsLabelLength = rhs.label.length();
+  return std::tie(lhs.found_, lhs.skip_, lhs.priority_, lhsFilterTextLength,
```
`!lhs.found_ `  ?

Sarcasm replied:
Thanks, good catch!

MaskRay commented on src/clang_complete.cc:
```diff
@@ -149,6 +150,103 @@ lsCompletionItemKind GetCompletionKind(CXCursorKind cursor_kind) {
+
+    std::string text;
+    switch (kind) {
+    case CXCompletionChunk_LeftParen:       text = '(';  break;
```
line 168 contains
```c++
    case CXCompletionChunk_LeftParen:       text = '(';  break;
    case CXCompletionChunk_RightParen:      text = ')';  break;
```
Can they be less duplicated

Sarcasm replied:
Not sure to understand, what do you mean?

MaskRay replied:
There is another chunk of code below where these character branches (`case CXCompletionChunk_LeftParen`) and string literals `text = '('` are duplicated

MaskRay commented:
Thx!

Merged in 16dd874af3c9dca4fe374ba8b4b92c0c89e28c7c
